### PR TITLE
Added regression_MWI_score function

### LINKED ISSUE #407: Add the Winkler Interval score
**Is your feature request related to a problem? Please describe.**
 I would like to calculate the Winkler Interval score. So that it can be used in the kaggle [Prediction interval competition I: Birth weight competition](https://www.kaggle.com/competitions/prediction-interval-competition-i-birth-weight/overview).

**Describe the solution you'd like** (link)[https://www.kaggle.com/datasets/carlmcbrideellis/winkler-interval-score-metric]
Model performance evaluation: The Mean Winkler Interval score (MWIS)
We can assess the overall performance of a regression model that produces prediction intervals by using the mean [Winkler Interval score](https://otexts.com/fpp3/distaccuracy.html#winkler-score) [1,2,3] which, for an individual interval, is given by:

image

where 𝑦 is the true value, 𝑢 it the upper prediction interval, 𝑙 is the lower prediction interval, and 𝛼 is (1-coverage). For example, for 90% coverage, 𝛼=0.1.

Note that the Winkler Interval score constitutes a proper scoring rule [2,3].

[1] [Robert L. Winkler "A Decision-Theoretic Approach to Interval Estimation", Journal of the American Statistical Association, volume 67, pp. 187-191 (1972)](https://doi.org/10.1080/01621459.1972.10481224)
[2] [Tilmann Gneiting and Adrian E Raftery "Strictly Proper Scoring Rules, Prediction, and Estimation", Journal of the American Statistical Association, volume 102, pp. 359-378 (2007)](https://doi.org/10.1198/016214506000001437) (Section 6.2)
[3] [Jonas R. Brehmer, and Tilmann Gneiting "Scoring interval forecasts: Equal-tailed, shortest, and modal interval", Bernoulli volume 27 pp. 1993-2010 (2021)](https://doi.org/10.3150/20-BEJ1298)

### PR DESCRIPTION
**Pull request by Carl-McBride-Ellis**
# Description

This is an efficient implementation for calculating the mean [Winkler interval score](https://otexts.com/fpp3/distaccuracy.html#winkler-score), based on the code written by Kaggle user [AmbrosM](https://www.kaggle.com/ambrosm) and published in the topic ["Accelerating the scoring function"](https://www.kaggle.com/competitions/prediction-interval-competition-i-birth-weight/discussion/461997).

Fixes #407

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

Private testing was performed in the Kaglge competition [Prediction interval competition I: Birth weight](https://www.kaggle.com/competitions/prediction-interval-competition-i-birth-weight/overview) by comparing correspondence with the results to those output by the competition metric code published in [Winkler Interval score metric](https://www.kaggle.com/datasets/carlmcbrideellis/winkler-interval-score-metric).

#### Update (2024-02-27)
A test routine for the function `regression_MWI_score` has been written and added to the end of `mapie/tests/test_metrics.py` ([b101cb0](https://github.com/scikit-learn-contrib/MAPIE/pull/408/commits/b101cb04a4bf9816da8a83ce458e9a74ed56bd04)).

# Checklist

- [x] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
- [x] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
- [x] Linting passes successfully : `make lint`
- [x] Typing passes successfully : `make type-check`
- [x] Unit tests pass successfully : `make tests`
- [x] Coverage is 100% : `make coverage`
- [x] Documentation builds successfully : `make doc`

### CODE REVIEW

**Review by LacombeLouis — APPROVED** [review]
Thank you @Carl-McBride-Ellis for this PR. We hope to soon make a release of this new metric so it can be used in the Kaggle competition! 

Note that we will need to add this change to the history and authors file.

**Comment by LacombeLouis** [comment]
We should definitely add some documentation, but we can do this in a separate PR, so this is pushed through as quickly as possible!

File: mapie/metrics.py [hunk 1]
```diff
@@ -1575,3 +1575,60 @@ def spiegelhalter_p_value(y_true: NDArray, y_score: NDArray) -> float:
...
+
+def regression_MWI_score(
+        y_true: NDArray,
+        y_pis: NDArray,
+        alpha: float
+) -> float:
+    """
+    This is an implementation of the Winkler interval score
+    (https://otexts.com/fpp3/distaccuracy.html#winkler-score).
+
```
Comment (candicemyt) [inline mapie/metrics.py:1588]: Could you add a short description of the score or the formula rather than a link ?

File: mapie/metrics.py [hunk 2]
```diff
@@ -1575,3 +1575,60 @@ def spiegelhalter_p_value(y_true: NDArray, y_score: NDArray) -> float:
...
+    # Checking for NaN and inf values
+    for array in (y_true, y_pred_low, y_pred_up):
+        check_array_nan(array)
+        check_array_inf(array)
+
+    width = np.sum(y_pred_up) - np.sum(y_pred_low)
+    error_above = np.sum((y_true - y_pred_up)[y_true > y_pred_up])
+    error_below = np.sum((y_pred_low - y_true)[y_true < y_pred_low])
+    total_error = error_above + error_below
+    MWIs = (width + total_error * 2 / alpha) / len(y_true)
```
Comment (candicemyt) [inline mapie/metrics.py:1633]: We would prefer variable names to be in lower case.

**Review by candicemyt — CHANGES_REQUESTED** [review]
Thank you @Carl-McBride-Ellis for this PR! It is something that we have wanted to add for a long time.
I have just added a few comments in order to be consistent with the rest of the library

File: mapie/metrics.py (hunk 2 above)
Comment (LacombeLouis) [reply mapie/metrics.py:1633]: done.

**Review by candicemyt — APPROVED** [review]

**Pull request merged**